diff --git a/include/ecrt.h b/include/ecrt.h
index 3ab462fb..ce22ef26 100644
--- a/include/ecrt.h
+++ b/include/ecrt.h
@@ -470,6 +470,8 @@ typedef struct {
     uint8_t sync_count; /**< Number of sync managers. */
     uint16_t sdo_count; /**< Number of SDOs. */
     char name[EC_MAX_STRING_LENGTH]; /**< Name of the slave. */
+    uint16_t error_code;
+    uint32_t abort_code;
 } ec_slave_info_t;
 
 /****************************************************************************/
diff --git a/lib/master.c b/lib/master.c
index 2a9018e5..03d8ff74 100644
--- a/lib/master.c
+++ b/lib/master.c
@@ -311,6 +311,9 @@ int ecrt_master_get_slave(ec_master_t *master, uint16_t slave_position,
     slave_info->sync_count = data.sync_count;
     slave_info->sdo_count = data.sdo_count;
     strncpy(slave_info->name, data.name, EC_MAX_STRING_LENGTH);
+
+    slave_info->error_code = data.error_code;
+    slave_info->abort_code = data.abort_code;
     return 0;
 }
 
diff --git a/master/fsm_change.c b/master/fsm_change.c
index 98fae184..b2ce0069 100644
--- a/master/fsm_change.c
+++ b/master/fsm_change.c
@@ -472,6 +472,9 @@ void ec_fsm_change_state_code(ec_fsm_change_t *fsm
                     al_msg->code, al_msg->message);
             break;
         }
+
+        fsm->slave->error_code = al_msg->code;
+
         if (al_msg->code == 0xffff) { /* not found in our list. */
             EC_SLAVE_ERR(fsm->slave, "Unknown AL status code 0x%04X.\n",
                     code);
diff --git a/master/fsm_coe.c b/master/fsm_coe.c
index 69765cd0..2a666cf4 100644
--- a/master/fsm_coe.c
+++ b/master/fsm_coe.c
@@ -60,7 +60,7 @@
 /****************************************************************************/
 
 // prototypes for private methods
-void ec_canopen_abort_msg(const ec_slave_t *, uint32_t);
+void ec_canopen_abort_msg(ec_slave_t *, uint32_t);
 int ec_fsm_coe_check_emergency(const ec_fsm_coe_t *, const uint8_t *, size_t);
 int ec_fsm_coe_prepare_dict(ec_fsm_coe_t *, ec_datagram_t *);
 int ec_fsm_coe_dict_prepare_desc(ec_fsm_coe_t *, ec_datagram_t *);
@@ -151,12 +151,14 @@ const ec_code_msg_t sdo_abort_messages[] = {
 /** Outputs an SDO abort message.
  */
 void ec_canopen_abort_msg(
-        const ec_slave_t *slave, /**< Slave. */
+        ec_slave_t *slave, /**< Slave. */
         uint32_t abort_code /**< Abort code to search for. */
         )
 {
     const ec_code_msg_t *abort_msg;
 
+    slave->abort_code = abort_code;
+
     for (abort_msg = sdo_abort_messages; abort_msg->code; abort_msg++) {
         if (abort_msg->code == abort_code) {
             EC_SLAVE_ERR(slave, "SDO abort message 0x%08X: \"%s\".\n",
diff --git a/master/ioctl.c b/master/ioctl.c
index 9a0e5c1b..66458c38 100644
--- a/master/ioctl.c
+++ b/master/ioctl.c
@@ -291,6 +291,9 @@ static ATTRIBUTES int ec_ioctl_slave(
     ec_ioctl_strcpy(data.order, slave->sii.order);
     ec_ioctl_strcpy(data.name, slave->sii.name);
 
+    data.error_code = slave->error_code;
+    data.abort_code = slave->abort_code;
+
     up(&master->master_sem);
 
     if (copy_to_user((void __user *) arg, &data, sizeof(data)))
diff --git a/master/ioctl.h b/master/ioctl.h
index 0c51e638..9388bc47 100644
--- a/master/ioctl.h
+++ b/master/ioctl.h
@@ -47,7 +47,7 @@
  *
  * Increment this when changing the ioctl interface!
  */
-#define EC_IOCTL_VERSION_MAGIC 37
+#define EC_IOCTL_VERSION_MAGIC (37 + (1 << 31))
 
 // Command-line tool
 #define EC_IOCTL_MODULE                EC_IOR(0x00, ec_ioctl_module_t)
@@ -260,6 +260,9 @@ typedef struct {
     char image[EC_IOCTL_STRING_SIZE];
     char order[EC_IOCTL_STRING_SIZE];
     char name[EC_IOCTL_STRING_SIZE];
+
+    uint16_t error_code;
+    uint32_t abort_code;
 } ec_ioctl_slave_t;
 
 /****************************************************************************/
diff --git a/master/slave.c b/master/slave.c
index 04cf2151..ea69b2a7 100644
--- a/master/slave.c
+++ b/master/slave.c
@@ -157,6 +157,9 @@ void ec_slave_init(
 
     // create state machine object
     ec_fsm_slave_init(&slave->fsm, slave);
+
+    slave->error_code = 0;
+    slave->abort_code = 0;
 }
 
 /****************************************************************************/
diff --git a/master/slave.h b/master/slave.h
index 4e113e28..1407fcfb 100644
--- a/master/slave.h
+++ b/master/slave.h
@@ -225,6 +225,8 @@ struct ec_slave
     struct list_head eoe_requests; /**< EoE set IP parameter requests. */
 
     ec_fsm_slave_t fsm; /**< Slave state machine. */
+    uint16_t error_code;
+    uint32_t abort_code;
 };
 
 /****************************************************************************/
